x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/aggs/InferencePipelineAggregationBuilder.java

```diff
@@ -0,0 +1,210 @@
+                                                            XContentParser parser) {
+        Tuple<SetOnce<ModelLoadingService>, String> context = new Tuple<>(modelLoadingService, pipelineAggregatorName);
+        return PARSER.apply(parser, context);
+    }
+
+    public InferencePipelineAggregationBuilder(String name, SetOnce<ModelLoadingService> modelLoadingService,
+                                               Map<String, String> bucketsPath) {
+        super(name, NAME, bucketsPath.values().toArray(new String[] {}));
+        this.modelLoadingService = modelLoadingService;
+        this.bucketPathMap = bucketsPath;
+    }
+
+    public InferencePipelineAggregationBuilder(StreamInput in, SetOnce<ModelLoadingService> modelLoadingService) throws IOException {
+        super(in, NAME);
+        modelId = in.readString();
+        bucketPathMap = in.readMap(StreamInput::readString, StreamInput::readString);
+        inferenceConfig = in.readOptionalNamedWriteable(InferenceConfig.class);
+        gapPolicy = BucketHelpers.GapPolicy.readFrom(in);
+        this.modelLoadingService = modelLoadingService;
+    }
+
+    void setModelId(String modelId) {
+        this.modelId = modelId;
+    }
+
+    void setInferenceConfig(InferenceConfig inferenceConfig) {
+        this.inferenceConfig = inferenceConfig;
+    }
+
+    void setGapPolicy(BucketHelpers.GapPolicy gapPolicy) {
+        this.gapPolicy = gapPolicy;
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+        context.validateHasParent(NAME, name);
+        if (modelId == null) {
+            context.addValidationError("Model Id must be set");
+        }
+        if (gapPolicy != BucketHelpers.GapPolicy.SKIP) {
+            context.addValidationError("gap policy [" + gapPolicy + "] in not valid for [" + NAME + "] aggregation");
+        }
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(modelId);
+        out.writeMap(bucketPathMap, StreamOutput::writeString, StreamOutput::writeString);
+        out.writeOptionalNamedWriteable(inferenceConfig);
+        gapPolicy.writeTo(out);
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metaData) {
+
+        SetOnce<Model> model = new SetOnce<>();
+        SetOnce<Exception> error = new SetOnce<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<Model> listener = new LatchedActionListener<>(
+            ActionListener.wrap(model::set, error::set), latch);
+
+        modelLoadingService.get().getModelForSearch(modelId, listener);
+        try {
+            // TODO Avoid the blocking wait
+            latch.await();
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Inference aggregation interrupted loading model", e);
+        }
+
+        if (error.get() != null) {
+            throw new RuntimeException(error.get());
```

Comment:
Most exceptions used in Elasticsearch are already `RuntimeException`s.  It would be better if these weren't wrapped in a generic runtime exception, as that will look less friendly in error messages and especially if it ends up getting transported to a different node.  So it would be better to only wrap the exception if it's not already `instanceof RuntimeException`.